crates/notion-core/src/shell/mod.rs

```diff
@@ -44,18 +37,12 @@ pub trait Shell {
 
 pub struct CurrentShell(Box<dyn Shell>);
 
-/// Thrown when the shell name was not specified in the Notion environment.
-#[derive(Debug, Fail, NotionFail)]
-#[fail(display = "Notion shell not specified")]
-#[notion_fail(code = "EnvironmentError")]
-struct UnspecifiedShellError;
-
 impl CurrentShell {
     pub fn detect() -> Fallible<Self> {
         match env::shell_name() {
             Some(name) => Ok(name.parse()?),
             None => {
-                throw!(UnspecifiedShellError);
+                throw!(ErrorDetails::UnspecifiedShell);
```

Comment:
Another place that the change made me notice—here gain the combinator approach yields at least a *different* style:


```rust
pub fn detect() -> Fallible<Self> {
    env::shell_name()
        .ok_or(ErrorDetails::UnspecifiedShell.into())
        .and_then(|name| name.parse())
}
```

Reply:
I'm not sure if this is particularly relevant, but one benefit to using the `match` approach is that the `ErrorDetails::UnspecifiedShell` object isn't created if there is no error. When using `.ok_or()`, we need to construct the error in order to pass the parameter. With the control flow from the `match` block, we only construct the error when we are about to return it.

We could get around that by using `.ok_or_else()`, but that would then create the closure function to pass as a parameter, so we're still constructing something.

All that said, this is almost assuredly premature optimization (and might not actually have any real benefits with compiler optimizations). Your formulation is a lot more readable, so I think until we notice major performance issues, going with readability is better :+1:

Reply:
Ah, yeah, that's definitely worth keeping in mind. One thing I keep reminding myself is that `match` is *good* – I have a bunch of deeply ingrained habits from living in JS/TS-land where there *is* no `match`, and am trying to remember: `match` is idiomatic! 😆